slides/D-trait-objects-patterns.md

```diff
@@ -0,0 +1,824 @@
+
+```rust
+const PATTERNS: &[Pattern] = &[
+    Pattern::new("Newtype"),
+    Pattern::new("RAII with guards"),
+    Pattern::new("Typestate"),
+    Pattern::new("Strategy"),
+];
+fn main() {
+    for pattern in PATTERNS {
+        pattern.introduce();
+        pattern.show_example();
+        pattern.when_to_use();
+    }
+}
+```
+
+---
+layout: statement
+---
+
+# 1. The Newtype pattern
+a small but useful pattern
+
+---
+layout: default
+---
+
+# Newtype: introduction
+&nbsp;
+
+### Wrap an external type in a new local type
+
+```rust
+pub struct Imei(String)
+```
+
+That's it!
+
+---
+layout: default
+---
+
+# Newtype: example
+
+```rust
+pub enum ValidateImeiError { /* - snip - */}
+
+pub struct Imei(String);
+
+impl Imei {
+    fn validate(imei: &str) -> Result<(), ValidateImeiError> {
+        todo!();
+    }
+}
+
+impl TryFrom<String> for Imei {
+    type Error = ValidateImeiError;
+
+    fn try_from(imei: String) -> Result<Self, Self::Error> {
+        Self::validate(&imei)?;
+        Ok(Self(imei))
+    }
+}
+
+fn register_phone(imei: Imei, label: String) {
+    // We can certain `imei` is valid here
+}
+```
+
+---
+layout: default
+---
+
+# Newtype: when to use
+
+Newtype solves some problems:
+- Orphan rule: no `impl`s for external `trait`s on external types
+- Allow for semantic typing (`url` example from mod B)
+- Enforce input validation
+
+---
+layout: statement
+---
+
+# 2. The RAII guard pattern
+More robust resource handling
+
+---
+layout: default
+---
+
+# RAII Guards: introduction
+
+- Resource Acquisition Is Initialization (?)
+- Couple resource aquisition and forfeiture with lifetime of a variable
+- Guard constructor initializes resource, destructor frees it
+- Access resource through the guard
+
+*Do you know of an example?*
+
+---
+layout: two-cols
+---
+
+# RAII Guards: example
+
+```rust
+pub struct Transaction<'c> {
+    connection: &'c mut Connection,
+    did_commit: bool,
+    id: usize,
+}
+
+impl<'c> Transaction<'c> {
+    pub fn begin(connection: &'c mut Connection)
+     -> Self {
+        let id = 
+            connection.start_transaction();
+        Self {
+            did_commit: false,
+            id,
+            connection,
+        }
+    }
+
+    pub fn query(&self sql: &str) { /* - snip - */}
+
+    pub fn commit(self) {
+        self.did_commit = true;
+    }
+}
+```
+::right::
+<div style="padding-left:10px; padding-top: 50px;">
+
+```rust
+impl Drop for Transaction<'_> {
+    fn drop(&mut self) {
+        if self.did_commit {
+            self
+                .connection
+                .commit_transaction(self.id);
+            
+        } else {
+            self
+                .connection
+                .rollback_transaction(self.id);
+        }
+    }
+}
+```
+</div>
+
+---
+layout: default
+---
+
+# RAII Guards: when to use
+
+- Ensure a resource is freed at some point
+- Ensure invariants hold while guard lives
+
+---
+layout: statement
+---
+
+# 3. The Typestate pattern
```

Comment:
When you get to this point, take stock of how much time you have left (and want to spend on things after this section). The typestate pattern is nice, but in principle people can also explore this in their own time (or during the project).

Reply:
I.e. you don't want to rush the "anti-patterns" because you lost the time in this section.